[BEAM-1887] Switch Apex ParDo to new DoFn.

tweise opened the pull request:
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [x] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [x] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [x] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [ ] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

---
R: @kennknowles @jkff 

This PR contains a subset of the changes required to support SplittableDoFn in the Apex runner:

* Switch ParDo operator from OldDoFn to DoFn
* Add AssignWindow primitive
* Fix serialization issues with StateInternals

kennknowles commented:
Run Apex ValidatesRunner

kennknowles commented:
retest this please

jkff commented on runners/apex/src/main/java/org/apache/beam/runners/apex/translation/operators/ApexGroupByKeyOperator.java:
```diff
@@ -95,7 +96,7 @@
   @Bind(JavaSerializer.class)
   private final StateInternalsFactory<K> stateInternalsFactory;
-  private Map<ByteBuffer, StateInternals<K>> perKeyStateInternals = new HashMap<>();
+  //private Map<ByteBuffer, StateInternals<K>> perKeyStateInternals = new HashMap<>();
```
Remove commented-out code?

tweise replied:
done

jkff commented on runners/apex/src/main/java/org/apache/beam/runners/apex/translation/operators/ApexParDoOperator.java:
```diff
@@ -83,6 +105,11 @@
   private final List<PCollectionView<?>> sideInputs;
 
+  private StateInternalsProxy<?> currentKeyStateInternals;
+  // TODO: fault tolerant timer internals
```
What are the implications of currently having non-fault tolerant ones? And what faults are these not tolerant to?

tweise replied:
The timer state is transient, if the operator gets restored to checkpointed state due to a failure, the timer state is lost. I'm going to add checkpointed timer state in the subsequent SDF PR.

tweise replied:
updated comment

jkff commented on runners/apex/src/main/java/org/apache/beam/runners/apex/translation/operators/ApexParDoOperator.java:
```diff
@@ -123,27 +151,15 @@ public ApexParDoOperator(
 
-  }
+    Coder<?> keyCoder = null;
+    if (doFn instanceof SplittableParDo.ProcessFn) {
```
This doesn't seem like it belongs to the current PR about porting to New DoFn

tweise replied:
removed

jkff commented on runners/apex/src/main/java/org/apache/beam/runners/apex/translation/operators/ApexParDoOperator.java:
```diff
@@ -305,31 +333,116 @@ public void setup(OperatorContext context) {
-    } catch (Exception e) {
-      Throwables.propagateIfPossible(e);
-      throw new RuntimeException(e);
+    if (doFn instanceof SplittableParDo.ProcessFn) {
```
Ditto

tweise replied:
removed

jkff commented on runners/apex/src/main/java/org/apache/beam/runners/apex/translation/operators/ApexProcessFnOperator.java:
```diff
@@ -0,0 +1,192 @@
+   */
+  public static class ProcessFn<InputT> implements Serializable {
+    public void process(ApexStreamTuple<WindowedValue<InputT>> input,
+        OutputEmitter<ApexStreamTuple<? extends WindowedValue<?>>> outputEmitter) throws Exception {
```
Shouldn't this be abstract then?

tweise replied:
done

tweise commented:
@jkff thanks for the review. I understand your concern about changes that are not strictly necessary being part of this PR. However, as you know this is not really a standalone change but part of an originally larger diff that aims to enable SplittableDoFn. It was discussed later to submit these changes separately and I already spent a good amount of time removing all the other splittable specific pieces. I think before spending even more time extracting the as you rightly point out unused code, I would rather return to the original plan and submit an even larger PR for the entire splittable feature. @kennknowles would like to have your opinion on that as well.

kennknowles commented on runners/apex/src/main/java/org/apache/beam/runners/apex/translation/operators/ApexParDoOperator.java:
```diff
@@ -393,4 +506,71 @@ public void clear() {
+    };
+
+    // TODO: fault tolerant timers
+    final Map<K, InMemoryTimerInternals> perKeyTimerInternals = new HashMap<>();
```
Is it certain that you have few enough keys for this to be safe?

tweise replied:
In the case of SDF I believe that is the case, in general not. I think this needs to be addressed before we can say that the runner supports timers.

jkff commented:
I am a bit conflicted here, because on one hand I'm very happy with progress on SDF and on removing the ugly remnants of OldDoFn, but on the other hand I'm not comfortable LGTMing code that I do not fully understand; plus there's concerns that Kenn expressed about clean rollbacks/cherrypicks, and general hygiene of file history.

For me, progress on the features in this case is sufficiently important that I also won't block on some of these concerns, but there are a few things that I think are simple enough to address, and would greatly improve the state of this PR, so I would prefer to see them addressed:

- As Kenn says, some of the large chunks of code here here seem easy to remove from the PR using git add -p. That would significantly reduce the size of this PR and make it easier to review.
- I'd really appreciate if you edited the PR description to explain the major changes you did here, and why. That also would make it easier to review for me (I would be able to review it much faster and with much higher quality); and I found that when I do this myself, I often find that there are things I missed, so it's good for code quality too.

tweise commented on runners/apex/src/main/java/org/apache/beam/runners/apex/translation/ApexPipelineTranslator.java:
```diff
@@ -174,4 +174,5 @@ public void translate(
       LOG.debug("view {}", view.getName());
     }
   }
+
```
fixed

tweise commented:
@kennknowles I made the changes and rebased as part of it. Do you see why the Apex build stopped after the unit tests?

tweise commented:
retest this please

tweise commented:
@kennknowles @jkff I addressed the comments and updated the PR description. BTW git add -p is very useful if the hunks are self contained, but not for cases where removal leaves behind unused import statements, for example. Overall, as much as I agree with the ideal situation, my time is also limited and I would like to spend it on those things that really benefit the Apex runner. I could have probably fixed a couple minor JIRAs instead of shifting hunks of code between PRs.

jkff commented on runners/apex/src/main/java/org/apache/beam/runners/apex/translation/operators/ApexParDoOperator.java:
```diff
@@ -25,41 +25,52 @@
 import org.apache.beam.runners.core.DoFnRunners;
 import org.apache.beam.runners.core.DoFnRunners.OutputManager;
 import org.apache.beam.runners.core.ExecutionContext;
-import org.apache.beam.runners.core.OldDoFn;
```
Woo! After this PR, the only usage of OldDoFn remains in ApexGroupByKeyOperator, and that one seems easy to remove too (not asking you to do it - just observing).

tweise replied:
Haha yes, I will create a JIRA for you to take that up :)

jkff commented on runners/apex/src/main/java/org/apache/beam/runners/apex/translation/operators/ApexParDoOperator.java:
```diff
@@ -255,6 +252,18 @@ public void process(ApexStreamTuple<WindowedValue<Iterable<?>>> t) {
       pushbackDoFnRunner.startBundle();
+      if (currentKeyStateInternals != null) {
+        InputT value = elem.getValue();
+        // TODO: generic key extraction
```
What is this referring to? I don't think there are other cases in Beam for extracting a key, except for KeyedWorkItem and KV.

tweise replied:
OK, if you think the current code is appropriate then I will remove the comment.

jkff commented on runners/apex/src/main/java/org/apache/beam/runners/apex/translation/operators/ApexParDoOperator.java:
```diff
@@ -83,6 +94,12 @@
   @Bind(JavaSerializer.class)
   private final List<PCollectionView<?>> sideInputs;
 
+  private StateInternalsProxy<?> currentKeyStateInternals;
```
Seems like this variable is never assigned to, so all branches testing for it being non-null are currently dead code. Suppose that's coming in a followup PR?

tweise replied:
Yes, this is result of shifting code between PRs.

jkff commented on runners/apex/src/main/java/org/apache/beam/runners/apex/translation/operators/ApexParDoOperator.java:
```diff
@@ -393,4 +441,71 @@ public void clear() {
+      public TimerInternals timerInternalsForKey(K key) {
+        InMemoryTimerInternals timerInternals = perKeyTimerInternals.get(key);
+        if (timerInternals == null) {
+          System.out.println("new timer internals for " + key);
```
Please remove system.out

jkff commented on runners/apex/src/main/java/org/apache/beam/runners/apex/translation/operators/ApexProcessFnOperator.java:
```diff
@@ -0,0 +1,191 @@
+   *
+   * @param <InputT>
+   */
+  public interface ProcessFn<InputT> extends Serializable {
```
I was very confused when reviewing this file, because this clashes with SplittableParDo.ProcessFn. Perhaps rename to something else?

tweise replied:
Yes, thought the same and could need a good idea. ApexOperatorFn?

jkff replied:
Yup, sounds good.

tweise commented:
@jkff I will make the changes later today and then I will prefer that this PR gets merged so that I can put back to the changes for SDF that are already implemented. As we discussed two days ago, the only thing missing was the timer support. And maybe I would be have been done with that by now!

jkff commented:
Run Apex ValidatesRunner

jkff commented:
LGTM, merging.
